### DISCUSSION TITLE
How to kick out of lists

### DISCUSSION THREAD

**Discussion by manunamz** [micromark/micromark#141] [Q&A]
Hi,

I recently [put out a couple of remark plugins](https://github.com/remarkjs/remark/compare/main...manunamz:remark:patch-1), but have a bug I haven't been able to wrap my head around.

In short, I am trying to get the target syntax -> `:attrtype::[[fname-a]]` to render properly when it appears after markdown lists, but the syntax should never render inside of markdown lists.

This is occurring in my [`micromark-extension-wikirefs`](https://github.com/wikibonsai/remark-wikirefs/tree/main/micromark-extension-wikirefs) plugin. [These](https://github.com/wikibonsai/remark-wikirefs/blob/main/micromark-extension-wikirefs/test/tests/runner.spec.ts#L149) are the failing tests whose test runner is [here](https://github.com/wikibonsai/remark-wikirefs/blob/main/micromark-extension-wikirefs/test/tests/runner.spec.ts#L21), full output you can checkout [here](https://gist.github.com/manunamz/bb046af2151131cc6efc9f0a9ae28bbe), and full test suite descriptions are [here](https://github.com/wikibonsai/wikirefs/tree/main/spec/cases).

I _think_ every single one of these failing tests stems from the fact that I'm not kicking out of markdown lists and I have not been able to figure out how to do this. I have tried [a few things](https://github.com/wikibonsai/remark-wikirefs/commit/54c9b32a7aa207a1cc0ca811aab3695b026419bc) to kick out of lists preemptively, but to no avail.

Here are some short examples:

Both of these markdown snippets...:

```markdown
- list-item-1
- list-item-2
- list-item-3

:attrtype::[[fname-a]]
```

```markdown
- list-item-1
- list-item-2
- list-item-3
:attrtype::[[fname-a]]
```

...HTML behavior:

```html
// desired
<aside class="attrbox">
<span class="attrbox-title">Attributes</span>
<dl>
<dt>attrtype</dt>
<dd><a class="attr wiki reftype__attrtype" href="/tests/fixtures/fname-a" data-href="/tests/fixtures/fname-a">title a</a></dd>
</dl>
</aside>
<ul>
<li>list-item-1</li>
<li>list-item-2</li>
<li>list-item-3</li>
</ul>

// actual
<ul>
<li>list-item-1</li>
<li>list-item-2</li>
<li><aside class="attrbox">
<span class="attrbox-title">Attributes</span>
<dl>
<dt>attrtype</dt>
<dd><a class="attr wiki reftype__attrtype" href="/tests/fixtures/fname-a" data-href="/tests/fixtures/fname-a">title a</a></dd>
</dl>
</aside>
list-item-3
</li>
</ul>
```

Any guidance in fixing these bugs would be much appreciated.

Note: I'm not sure if the [lazy line discussion](https://github.com/micromark/micromark/discussions/127) applies here. I also looked at [this discussion](https://github.com/micromark/micromark/discussions/58), but still feel a tad lost.

**Comment by wooorm** [comment]
Hey!

Importantly, why not use directives, inventing yet another syntax means your markdown can’t be ported, it‘s going to be a bad experience over the years!

Or, something that exists, such as https://github.com/landakram/micromark-extension-wiki-link?

> the syntax should never render inside of markdown lists.

What about folks writing `- :my-extension`?

> [These](https://github.com/wikibonsai/remark-wikirefs/blob/main/micromark-extension-wikirefs/test/tests/runner.spec.ts#L149) are the failing tests whose test runner is [here](https://github.com/wikibonsai/remark-wikirefs/blob/main/micromark-extension-wikirefs/test/tests/runner.spec.ts#L21), full output you can checkout [here](https://gist.github.com/manunamz/bb046af2151131cc6efc9f0a9ae28bbe), and full test suite descriptions are [here](https://github.com/wikibonsai/wikirefs/tree/main/spec/cases).

Thanks for sharing so much in detail, but that’s a **lot** for me to wade through, to get your entire project into my head!

You have four syntax extensions in your package. Which is the one with problems? All?

My first guess is that you are not `nok`ing on lazy lines. That’s a lazy line in a list item. Which you don’t want.
Here’s an example that calls `nok` if the current line is lazy: https://github.com/micromark/micromark-extension-gfm-table/blob/fba8bc7d597af0bc86552eb15b00b8d1be296d31/dev/lib/syntax.js#L576.

**Reply by manunamz** [reply]
Hello!

#### tldr;

re:
> My first guess is that you are not noking on lazy lines. That’s a lazy line in a list item. Which you don’t want.
Here’s an example that calls nok if the current line is lazy: https://github.com/micromark/micromark-extension-gfm-table/blob/fba8bc7d597af0bc86552eb15b00b8d1be296d31/dev/lib/syntax.js#L576.

Yes -- I was wondering about that! I [tried it here](https://github.com/wikibonsai/remark-wikirefs/blob/main/micromark-extension-wikirefs/src/lib/syntax-wikiattr.ts#L265), but upon another inspection I think I was mistaken about the problem being in the syntax parser. I think it is actually my [resolver](https://github.com/wikibonsai/remark-wikirefs/blob/main/micromark-extension-wikirefs/src/lib/syntax-wikiattr.ts#L162-L194).

The resolver collects `wikiattr` tokens and renders them at the top of the file. The result should feel similar to footnote definitions (which are collected throughout a markdown document and then rendered at the bottom of the page).

I think what's happening is that once I'm inside a list, the resolver moves the `wikiattr` to the "top" of the current list item instead of moving the token(s) to the top of the entire document. I am now struggling to figure out **how to move out of a subtokenizer in order to place those tokens at the top of the `document` level in my resolver.**

---

re: 
> Importantly, why not use directives, inventing yet another syntax means your markdown can’t be ported, it‘s going to be a bad experience over the years!

I'm so glad you asked! `[[Wikilinks]]` have become quite popular in the PKM (personal knowledge management) space and the wiki syntax has been expanded quite a bit, with some variation from app to app. [remark-wikirefs](https://github.com/wikibonsai/remark-wikirefs) is a remark plugin that implements an extended wiki syntax as detailed in [wikirefs-spec](https://github.com/wikibonsai/wikirefs/tree/main/spec), which tries to capture the most widely used wiki syntaxes. The hope is to encourage PKM apps to standardize on **something** so that more and more markdown with this syntax can be ported universally! **This is already happening**, but by providing open source solutions it makes it easier for users and developers to just turn on the syntax -- especially in markdown editors where PKM was not the first priority. 🔌 ✅

re:
> Or, something that exists, such as https://github.com/landakram/micromark-extension-wiki-link?

I looked at that plugin. It's a nice project!  It is also missing some syntaxes that PKM apps tend to support these days.

re:
> What about folks writing - :my-extension?

Is this supposed to be a markdown list item that begins with a colon? I'm not sure what you're asking here. 🤔

re:
> Thanks for sharing so much in detail, but that’s a lot for me to wade through, to get your entire project into my head!

My apologies! Just trying to figure out how much context is useful for other maintainers -- sorry for the fire-hose! 🚿

re:
> You have four syntax extensions in your package. Which is the one with problems? All?

There is only a problem with the [`wikiattr`](https://github.com/wikibonsai/remark-wikirefs/blob/main/micromark-extension-wikirefs/src/lib/syntax-wikiattr.ts) construct and only when the syntax occurs after markdown lists as far as I know -- everything else works!

---

And thank you for your time and timely attention! 🙏

**Reply by wooorm** [reply]
I don’t think you should use resolvers to move things around like that? Definitions or footnote definitions don’t do that either?

---

> [remark-wikirefs](https://github.com/wikibonsai/remark-wikirefs) is a remark plugin that implements an extended wiki syntax as detailed in [wikirefs-spec](https://github.com/wikibonsai/wikirefs/tree/main/spec), which tries to capture the most widely used wiki syntaxes. The hope is to encourage PKM apps to standardize on something

Doesn’t creating another standard defeat that purpose?

> It is also missing some syntaxes that PKM apps tend to support these days.

Maybe try working together? I dunno!

I also see wikilinks being used more and more. Perhaps at some point it makes sense to support them here too.
But I prefer directives. One solution to solve them all! :)

> > What about folks writing `- :my-extension`?

> Is this supposed to be a markdown list item that begins with a colon? I'm not sure what you're asking here. 🤔

A list, with an extension in it. To put it differently: `- a`, where `a` is your wikilink or whatever you want to support!

---

But looks like you have some ideas now on how to continue? By not moving in resolvers, by not moving to the start of lists?
